packages/vue-virtual/src/index.ts

```diff
@@ -0,0 +1,96 @@
+import {
+  elementScroll,
+  observeElementOffset,
+  observeElementRect,
+  observeWindowOffset,
+  observeWindowRect,
+  Virtualizer,
+  windowScroll,
+  type PartialKeys,
+  type VirtualizerOptions,
+} from '@tanstack/virtual-core'
+import {
+  computed,
+  nextTick,
+  shallowRef,
+  triggerRef,
+  unref,
+  watchEffect,
+  type Ref,
+} from 'vue'
+
+type MaybeRef<T> = T | Ref<T>
+
+function useVirtualizerBase<TScrollElement, TItemElement = unknown>(
+  options: MaybeRef<VirtualizerOptions<TScrollElement, TItemElement>>,
+): Ref<Virtualizer<TScrollElement, TItemElement>> {
+  const virtualizer = new Virtualizer(unref(options))
+  const state = shallowRef(virtualizer)
+
+  watchEffect((onCleanup) => {
+    const opts = unref(options)
+
+    // If those functions use refs, we need to track them.
+    opts.getScrollElement()
+    opts.estimateSize(0)
+
+    // We don't want to track state here
+    nextTick(() => {
+      virtualizer.setOptions({
+        ...opts,
+        onChange: (instance: Virtualizer<TScrollElement, TItemElement>) => {
+          // Force an update event
+          triggerRef(state)
+          opts.onChange?.(instance)
+        },
+      })
+
+      virtualizer._willUpdate()
```

Comment:
Adding another `triggerRef(state)` at this level fixed the dom update for my use case in my app but does not for the codesandbox example that I provided above